File: sdk/demo/python/src/prohand_demo/kapandji.py
#!/usr/bin/env python3
"""
ProHand SDK FFI Demo: Kapandji Opposition Test

Runs the Kapandji opposition sequence - thumb touches each fingertip.
"""

import sys
import argparse
import time
import math
from pathlib import Path
from typing import Dict, List, Optional
from .utils import DemoBase


def _safe_yaml_load(path: str) -> Optional[dict]:
    """Safely load YAML configuration file."""
    try:
        import yaml
    except ImportError:
        print("PyYAML not installed; install it to use YAML gestures.")
        return None
    try:
        with open(path, "r", encoding="utf-8") as f:
            return yaml.safe_load(f)
    except Exception as e:
        print(f"Failed to load YAML config from {path}: {e}")
        return None


def _parse_joint_list(values: Optional[List[float]], n: int, fill: float = 0.0) -> List[float]:
    """Parse joint list from YAML, filling missing values."""
    if values is None:
        return [float(fill)] * n
    out = [float(v) for v in values][:n]
    if len(out) < n:
        out.extend([float(fill)] * (n - len(out)))
    return out


def _pose_from_yaml(gesture: str, hand: str, cfg: dict) -> Dict[str, List[float]]:
    """Extract pose from YAML configuration."""
    hands = (cfg or {}).get("hands", {})
    if not hands:
        raise ValueError("YAML config missing 'hands' section")
    g = hands.get(gesture)
    if g is None:
        available = ", ".join(sorted(hands.keys()))
        raise ValueError(f"YAML gesture '{gesture}' not found. Available: {available}")

    def deg_to_rad_list(values: List[float], n: int) -> List[float]:
        vals = _parse_joint_list(values, n, fill=0.0)
        return [math.radians(v) for v in vals]

    thumb = deg_to_rad_list(g.get("thumb", []), 4)
    index = deg_to_rad_list(g.get("index", []), 4)
    middle = deg_to_rad_list(g.get("middle", []), 4)
    ring = deg_to_rad_list(g.get("ring", []), 4)
    pinky = deg_to_rad_list(g.get("pinky", []), 4)
    wrist = deg_to_rad_list(g.get("wrist", []), 2) if "wrist" in g else [0.0, 0.0]
    return {"thumb": thumb, "index": index, "middle": middle, "ring": ring, "pinky": pinky, "wrist": wrist}


def _pose_to_hand_positions(pose: Dict[str, List[float]]) -> List[float]:
    """Convert pose dictionary to flat list of 20 joint positions."""
    # Order: thumb[0-3], index[4-7], middle[8-11], ring[12-15], pinky[16-19]
    positions = []
    positions.extend(pose["thumb"])
    positions.extend(pose["index"])
    positions.extend(pose["middle"])
    positions.extend(pose["ring"])
    positions.extend(pose["pinky"])
    return positions


def _stream_pose(client, pose: Dict[str, List[float]], publish_hz: float, duration_s: float, torque: float = 0.45) -> None:
    """Stream a pose for the specified duration."""
    period = 1.0 / max(1e-6, float(publish_hz))
    positions = _pose_to_hand_positions(pose)
    wrist_positions = pose.get("wrist", [0.0, 0.0])
    
    deadline = time.time() + float(duration_s)
    while time.time() < deadline:
        client.send_wrist_streams(wrist_positions)
        client.send_hand_streams(positions, torque)
        time.sleep(period)


class KapandjiDemo(DemoBase):
    """Kapandji opposition test demo."""
    
    def __init__(self):
        super().__init__("ProHand Kapandji Opposition Test")
    
    def run(
        self,
        command_endpoint: str,
        status_endpoint: str,
        hand_streaming_endpoint: str,
        wrist_streaming_endpoint: str,
        yaml_config: str,
        hand: str,
        publish_freq: float
    ) -> int:
        """Run the Kapandji demo."""
        self.banner()
        
        print("\nConfiguration:")
        print(f"  YAML config:      {yaml_config}")
        print(f"  Hand:             {hand}")
        print(f"  Publish rate:     {publish_freq} Hz")
        
        client = None
        try:
            # Load YAML configuration
            self.section("Loading YAML configuration...")
            cfg = _safe_yaml_load(yaml_config)
            if cfg is None:
                self.error("YAML config required for Kapandji sequence")
                return 1
            self.success("YAML configuration loaded!")
            
            # Resolve torque level from YAML
            torque_level: float = 0.45
            try:
                level = (cfg or {}).get("default_torque_level")
                tmap = (cfg or {}).get("torque_map", {}) or {}
                if isinstance(level, str) and level in tmap:
                    torque_level = float(tmap[level])
            except Exception:
                torque_level = 0.45
            
            # Connect to IPC host
            self.section("Connecting to IPC host with streaming...")
            client = self.sdk.ProHandClient(
                command_endpoint, status_endpoint, hand_streaming_endpoint, wrist_streaming_endpoint
            )
            self.success("Connected with streaming support!")
            
            # Verify connection
            self.section("Verifying connection...")
            client.send_ping()
            time.sleep(0.2)
            self.success("Connection verified!")
            
            # Enable streaming mode
            self.section("Enabling streaming mode...")
            client.set_streaming_mode(True)
            if not client.wait_for_streaming_ready(timeout=10.0):
                self.error("Streaming connection failed to establish!")
                return 1
            self.success("Streaming mode enabled!")
            
            # Run Kapandji sequence
            self.section("Running Kapandji opposition sequence...")
            sequence: List[tuple[str, float]] = [
                # Slow
                ("finger_down_0", 2.0),
                ("finger_down_1", 1.0),
                ("finger_down_2", 1.0),
                ("finger_down_3", 1.0),
                ("finger_down_4", 1.0),
                ("finger_down_3", 1.0),
                ("finger_down_2", 1.0),
                ("finger_down_1", 1.0),
                
                # Medium
                ("finger_down_0", 1.25),
                ("finger_down_1", 0.5),
                ("finger_down_2", 0.5),
                ("finger_down_3", 0.5),
                ("finger_down_4", 0.5),
                ("finger_down_3", 0.5),
                ("finger_down_2", 0.5),
                ("finger_down_1", 0.5),
                
                # Fast
                ("finger_down_0", 0.5),
                ("finger_down_1", 0.25),
                ("finger_down_2", 0.25),
                ("finger_down_3", 0.25),
                ("finger_down_4", 0.25),
                ("finger_down_3", 0.25),
                ("finger_down_2", 0.25),
                ("finger_down_1", 0.25),
                
                # Fastest
                ("finger_down_0", 0.25),
                ("finger_down_1", 0.1),
                ("finger_down_2", 0.1),
                ("finger_down_3", 0.1),
                ("finger_down_4", 0.1),
                ("finger_down_3", 0.1),
                ("finger_down_2", 0.1),
                ("finger_down_1", 0.1),
                
                # Fastest (repeat)
                ("finger_down_0", 0.25),
                ("finger_down_1", 0.1),
                ("finger_down_2", 0.1),
                ("finger_down_3", 0.1),
                ("finger_down_4", 0.1),
                ("finger_down_3", 0.1),
                ("finger_down_2", 0.1),
                ("finger_down_1", 0.1),
                
                # Fastest (repeat)
                ("finger_down_0", 0.25),
                ("finger_down_1", 0.1),
                ("finger_down_2", 0.1),
                ("finger_down_3", 0.1),
                ("finger_down_4", 0.1),
                ("finger_down_3", 0.1),
                ("finger_down_2", 0.1),
                ("finger_down_1", 0.1),
            ]
            
            for gesture, dur in sequence:
                pose = _pose_from_yaml(gesture, hand, cfg)
                print(f"  Gesture: {gesture} for {dur:.2f}s")
                _stream_pose(client, pose, publish_freq, dur, torque_level)
            
            # Return to zero
            self.section("Returning to zero position...")
            zero_positions = [0.0] * 20
            client.send_hand_streams(zero_positions, torque_level)
            client.send_wrist_streams([0.0, 0.0])
            time.sleep(0.5)
            
            # Disable streaming mode
            self.section("Disabling streaming mode...")
            client.set_streaming_mode(False)
            time.sleep(0.2)
            
            self.success("Kapandji sequence completed!")
            return 0
            
        except self.sdk.ConnectionError as e:
            self.error(f"Connection failed: {e}")
            print("\nMake sure prohand-headless-ipc-host is running")
            return 1
        except self.sdk.ProHandError as e:
            self.error(f"Kapandji test failed: {e}")
            return 1
        except Exception as e:
            self.error(f"Unexpected error: {e}")
            import traceback
            traceback.print_exc()
            return 1
        finally:
            if client is not None:
                try:
                    client.set_streaming_mode(False)
                except Exception:
                    pass
                try:
                    client.close()
                except Exception:
                    pass


def main() -> int:
    """Main entry point."""
    parser = argparse.ArgumentParser(
        description="Kapandji opposition sequence (placeholder)",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )
    parser.add_argument(
        "--command-endpoint",
        type=str,
        default="ipc:///tmp/prohand-commands.ipc",
        help="ZMQ command endpoint"
    )
    parser.add_argument(
        "--status-endpoint",
        type=str,
        default="ipc:///tmp/prohand-status.ipc",
        help="ZMQ status endpoint"
    )
    parser.add_argument(
        "--hand-streaming-endpoint",
        type=str,
        default="ipc:///tmp/prohand-hand-streaming.ipc",
        help="ZMQ hand streaming endpoint",
    )
    parser.add_argument(
        "--wrist-streaming-endpoint",
        type=str,
        default="ipc:///tmp/prohand-wrist-streaming.ipc",
        help="ZMQ wrist streaming endpoint",
    )
    # Default YAML path relative to this script's location
    script_dir = Path(__file__).parent
    default_yaml = script_dir / "../../../config/kapandji.yaml"
    default_yaml = default_yaml.resolve()
    
    parser.add_argument(
        "--yaml-config",
        type=str,
        default=str(default_yaml),
        help="Path to YAML configuration file"
    )
    parser.add_argument(
        "--hand",
        type=str,
        default="left",
        choices=["left", "right"],
        help="Which hand configuration to use"
    )
    parser.add_argument(
        "--publish-frequency",
        type=float,
        default=60.0,
        help="Command publish rate (Hz)"
    )
    args = parser.parse_args()
    
    demo = KapandjiDemo()
    return demo.run(
        args.command_endpoint,
        args.status_endpoint,
        args.hand_streaming_endpoint,
        args.wrist_streaming_endpoint,
        args.yaml_config,
        args.hand,
        args.publish_frequency
    )


if __name__ == "__main__":
    sys.exit(main())

File: sdk/demo/cpp/src/kapandji.cpp
/**
 * ProHand SDK FFI Demo: Kapandji Opposition Test
 * 
 * Run the Kapandji opposition sequence - thumb touches each fingertip.
 */

#include <iostream>
#include <cxxopts.hpp>
#include <chrono>
#include <thread>
#include <cmath>
#include <vector>
#include <map>
#include <string>
#include <fstream>

#include <prohand_sdk/ProHandClient.hpp>
#include "prohand_demo/Utils.hpp"

#ifdef YAML_CPP_FOUND
#include <yaml-cpp/yaml.h>
#endif

using namespace prohand_sdk;
using namespace prohand_demo;  // For utils namespace

#ifdef YAML_CPP_FOUND
// Helper to parse joint list from YAML
std::vector<double> parseJointList(const YAML::Node& node, int n, double fill = 0.0) {
    std::vector<double> out(n, fill);
    if (node && node.IsSequence()) {
        for (int i = 0; i < n && i < static_cast<int>(node.size()); ++i) {
            out[i] = node[i].as<double>() * M_PI / 180.0;  // Convert degrees to radians
        }
    }
    return out;
}

// Extract pose from YAML configuration
std::map<std::string, std::vector<double>> poseFromYaml(
    const std::string& gesture, 
    const std::string& hand, 
    const YAML::Node& cfg) 
{
    auto hands = cfg["hands"];
    
    if (!hands) {
        throw std::runtime_error("YAML config missing 'hands' section");
    }
    
    auto g = hands[gesture];
    if (!g) {
        throw std::runtime_error("YAML gesture '" + gesture + "' not found");
    }
    
    std::map<std::string, std::vector<double>> pose;
    pose["thumb"] = parseJointList(g["thumb"], 4);
    pose["index"] = parseJointList(g["index"], 4);
    pose["middle"] = parseJointList(g["middle"], 4);
    pose["ring"] = parseJointList(g["ring"], 4);
    pose["pinky"] = parseJointList(g["pinky"], 4);
    
    auto wrist = parseJointList(g["wrist"], 2);
    if (wrist.size() < 2) wrist = {0.0, 0.0};
    pose["wrist"] = wrist;
    
    return pose;
}

// Convert pose dictionary to flat list of 20 joint positions
std::vector<float> poseToHandPositions(const std::map<std::string, std::vector<double>>& pose) {
    std::vector<float> positions;
    positions.reserve(20);
    
    // Order: thumb[0-3], index[4-7], middle[8-11], ring[12-15], pinky[16-19]
    const std::vector<std::string> fingerNames = {"thumb", "index", "middle", "ring", "pinky"};
    for (const auto& finger : fingerNames) {
        auto it = pose.find(finger);
        if (it != pose.end()) {
            for (double val : it->second) {
                positions.push_back(static_cast<float>(val));
            }
        } else {
            positions.insert(positions.end(), 4, 0.0f);
        }
    }
    
    return positions;
}

// Stream a pose for the specified duration
void streamPose(
    ProHandClient& client,
    const std::map<std::string, std::vector<double>>& pose,
    double publishHz,
    double durationS,
    float torque = 0.45f) 
{
    double period = 1.0 / std::max(1e-6, publishHz);
    std::vector<float> positions = poseToHandPositions(pose);
    auto wristIt = pose.find("wrist");
    std::vector<float> wristPositions = {0.0f, 0.0f};
    if (wristIt != pose.end()) {
        for (size_t i = 0; i < wristIt->second.size() && i < 2; ++i) {
            wristPositions[i] = static_cast<float>(wristIt->second[i]);
        }
    }
    
    auto deadline = std::chrono::steady_clock::now() + 
                    std::chrono::duration<double>(durationS);
    
    while (std::chrono::steady_clock::now() < deadline) {
        client.sendWristStreams(wristPositions);
        client.sendHandStreams(positions, torque);
        std::this_thread::sleep_for(std::chrono::duration<double>(period));
    }
}
#endif

int main(int argc, char** argv) {
    cxxopts::Options options("kapandji", "Kapandji opposition sequence");
    options.add_options()
        ("command-endpoint", "ZMQ command endpoint",
         cxxopts::value<std::string>()->default_value("ipc:///tmp/prohand-commands.ipc"))
        ("status-endpoint", "ZMQ status endpoint",
         cxxopts::value<std::string>()->default_value("ipc:///tmp/prohand-status.ipc"))
        ("hand-streaming-endpoint", "ZMQ hand streaming endpoint",
         cxxopts::value<std::string>()->default_value("ipc:///tmp/prohand-hand-streaming.ipc"))
        ("wrist-streaming-endpoint", "ZMQ wrist streaming endpoint",
         cxxopts::value<std::string>()->default_value("ipc:///tmp/prohand-wrist-streaming.ipc"))
        ("yaml-config", "Path to YAML configuration file",
         cxxopts::value<std::string>()->default_value("../config/kapandji.yaml"))
        ("hand", "Which hand configuration to use",
         cxxopts::value<std::string>()->default_value("left"))
        ("publish-frequency", "Command publish rate (Hz)",
         cxxopts::value<double>()->default_value("60.0"))
        ("h,help", "Print usage");
    
    auto result = options.parse(argc, argv);
    
    if (result.count("help")) {
        std::cout << options.help() << std::endl;
        return 0;
    }
    
    std::string commandEndpoint = result["command-endpoint"].as<std::string>();
    std::string statusEndpoint = result["status-endpoint"].as<std::string>();
    std::string handStreamingEndpoint = result["hand-streaming-endpoint"].as<std::string>();
    std::string wristStreamingEndpoint = result["wrist-streaming-endpoint"].as<std::string>();
    std::string yamlConfig = result["yaml-config"].as<std::string>();
    std::string hand = result["hand"].as<std::string>();
    double publishFreq = result["publish-frequency"].as<double>();
    
    utils::printBanner("ProHand Kapandji Opposition Test");
    std::cout << "\nConfiguration:\n";
    std::cout << "  YAML config:      " << yamlConfig << "\n";
    std::cout << "  Hand:             " << hand << "\n";
    std::cout << "  Publish rate:     " << publishFreq << " Hz\n";
    
#ifdef YAML_CPP_FOUND
    try {
        // Load YAML configuration
        utils::printSection("Loading YAML configuration...");
        YAML::Node cfg = YAML::LoadFile(yamlConfig);
        utils::printSuccess("YAML configuration loaded!");
        
        // Resolve torque level from YAML
        float torqueLevel = 0.45f;
        try {
            if (cfg["default_torque_level"]) {
                std::string level = cfg["default_torque_level"].as<std::string>();
                if (cfg["torque_map"] && cfg["torque_map"][level]) {
                    torqueLevel = static_cast<float>(cfg["torque_map"][level].as<double>());
                }
            }
        } catch (...) {
            torqueLevel = 0.45f;  // Keep default on any parse error
        }
        
        // Connect to IPC host
        utils::printSection("Connecting to IPC host with streaming...");
        ProHandClient client(commandEndpoint, statusEndpoint, handStreamingEndpoint, wristStreamingEndpoint);
        utils::printSuccess("Connected with streaming support!");
        
        // Verify connection
        utils::printSection("Verifying connection...");
        client.sendPing();
        std::this_thread::sleep_for(std::chrono::milliseconds(200));
        utils::printSuccess("Connection verified!");
        
        // Enable streaming mode
        utils::printSection("Enabling streaming mode...");
        utils::printInfo("Telling driver to accept streaming commands...");
        client.setStreamingMode(true);
        
        utils::printInfo("Waiting for streaming connection to be ready...");
        if (!client.waitForStreamingReady(10.0)) {
            utils::printError("Streaming connection failed to establish!");
            std::cout << "\nMake sure:\n";
            std::cout << "  1. prohand-headless-ipc-host is running\n";
            std::cout << "  2. Driver has streaming endpoint enabled\n";
            std::cout << "  3. All ZMQ endpoints match between client and driver\n";
            return 1;
        }
        utils::printSuccess("Streaming mode enabled!");
        
        // Run Kapandji sequence
        utils::printSection("Running Kapandji opposition sequence...");
        std::vector<std::pair<std::string, double>> sequence = {
            // Slow
            {"finger_down_0", 2.0},
            {"finger_down_1", 1.0},
            {"finger_down_2", 1.0},
            {"finger_down_3", 1.0},
            {"finger_down_4", 1.0},
            {"finger_down_3", 1.0},
            {"finger_down_2", 1.0},
            {"finger_down_1", 1.0},
            
            // Medium
            {"finger_down_0", 1.25},
            {"finger_down_1", 0.5},
            {"finger_down_2", 0.5},
            {"finger_down_3", 0.5},
            {"finger_down_4", 0.5},
            {"finger_down_3", 0.5},
            {"finger_down_2", 0.5},
            {"finger_down_1", 0.5},
            
            // Fast
            {"finger_down_0", 0.5},
            {"finger_down_1", 0.25},
            {"finger_down_2", 0.25},
            {"finger_down_3", 0.25},
            {"finger_down_4", 0.25},
            {"finger_down_3", 0.25},
            {"finger_down_2", 0.25},
            {"finger_down_1", 0.25},
            
            // Fastest
            {"finger_down_0", 0.25},
            {"finger_down_1", 0.1},
            {"finger_down_2", 0.1},
            {"finger_down_3", 0.1},
            {"finger_down_4", 0.1},
            {"finger_down_3", 0.1},
            {"finger_down_2", 0.1},
            {"finger_down_1", 0.1},
            
            // Fastest (repeat)
            {"finger_down_0", 0.25},
            {"finger_down_1", 0.1},
            {"finger_down_2", 0.1},
            {"finger_down_3", 0.1},
            {"finger_down_4", 0.1},
            {"finger_down_3", 0.1},
            {"finger_down_2", 0.1},
            {"finger_down_1", 0.1},
            
            // Fastest (repeat)
            {"finger_down_0", 0.25},
            {"finger_down_1", 0.1},
            {"finger_down_2", 0.1},
            {"finger_down_3", 0.1},
            {"finger_down_4", 0.1},
            {"finger_down_3", 0.1},
            {"finger_down_2", 0.1},
            {"finger_down_1", 0.1},
        };
        
        for (const auto& [gesture, dur] : sequence) {
            auto pose = poseFromYaml(gesture, hand, cfg);
            std::cout << "  Gesture: " << gesture << " for " << dur << "s\n";
            streamPose(client, pose, publishFreq, dur, torqueLevel);
        }
        
        // Return to zero
        utils::printSection("Returning to zero position...");
        std::vector<float> zeroPositions(20, 0.0f);
        client.sendHandStreams(zeroPositions, torqueLevel);
        std::this_thread::sleep_for(std::chrono::milliseconds(500));
        
        // Disable streaming mode
        utils::printSection("Disabling streaming mode...");
        client.setStreamingMode(false);
        std::this_thread::sleep_for(std::chrono::milliseconds(200));
        
        utils::printSuccess("Kapandji sequence completed!");
        return 0;
        
    } catch (const SdkException& e) {
        utils::printError(std::string("Kapandji test failed: ") + e.what());
        return 1;
    } catch (const std::exception& e) {
        utils::printError(std::string("Unexpected error: ") + e.what());
        return 1;
    }
#else
    utils::printError("YAML support not available (yaml-cpp not found)");
    std::cout << "\n⚠️  NOTE: This demo requires yaml-cpp to be installed.\n";
    std::cout << "    Install yaml-cpp and rebuild to enable full functionality.\n";
    std::cout << "    For now, this is a placeholder demo.\n\n";
    
    try {
        utils::printSection("Connecting to IPC host...");
        ProHandClient client(commandEndpoint, statusEndpoint, handStreamingEndpoint, wristStreamingEndpoint);
        utils::printSuccess("Connected!");
        
        std::cout << "\nIntended Kapandji sequence:\n";
        std::cout << "  1. Enable streaming mode\n";
        std::cout << "  2. Load positions from " << yamlConfig << "\n";
        std::cout << "  3. Run opposition sequence with varying speeds\n";
        std::cout << "  4. Disable streaming mode\n";
        
        return 0;
    } catch (const SdkException& e) {
        utils::printError(std::string("Connection failed: ") + e.what());
        return 1;
    }
#endif
}
